### PR TITLE
Remove reliance on npy_<impl> ufunc loops.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This removes Numba's reliance on NumPy ufunc inner loops. All the cases where a ufunc inner loop (or something from `numpy/npy_math.h`) was present, have been replaced as follows:

* `np.signbit` -> Numba custom builder implementation.
* `np.nextafter` -> Numba helperlib wrapped C library call.
* `np.spacing` -> Numba custom builder implementation.
* `np.log2` -> llvm intrinsic.
* `np.exp2` -> llvm intrinsic.
* `np.logaddexp` -> custom Numba overload translation of NumPy C code.
* `np.logaddexp2` -> custom Numba overload translation of NumPy C code.
* Numba helperlib `numba_cpowf` -> now uses `<complex.h>` functions in place of NumPy functions.

Edit: This also:
* Removes static linkage to `libnpymath.a` and replaces use of `numpy.distutils.misc_util`  with `np.get_include()` as there's now no for link information.

### CODE REVIEW

File: numba/np/npyfuncs.py [hunk 1]
```diff
@@ -1590,23 +1638,39 @@ def np_real_nextafter_impl(context, builder, sig, args):
...
     return _dispatch_func_by_name_type(context, builder, sig, args,
                                        dispatch_table, 'nextafter')
 
 def np_real_spacing_impl(context, builder, sig, args):
+    # This is different to how NumPy does it, NumPy has a specialisation of
+    # nextafter called _next, which is used. See:
+    # https://github.com/numpy/numpy/blob/12c2b7dd62fc0c14b81c8892ed5f4f59cc94d09c/numpy/core/src/npymath/ieee754.c.src#L32-L38
+    # Numba elects to use `nextafter` for a similar behaviour to save
+    # translating this very involved function. Further, the NumPy comments note
+    # that there is a lot of redundancy present betwen the two.
```
Comment (stuartarchibald) [inline numba/np/npyfuncs.py:1654]:
```suggestion
    # that there is a lot of redundancy present between the two.
```

File: numba/_helperlib.c [hunk 2]
```diff
@@ -8,17 +8,21 @@
 #include <stddef.h>
 #include <stdio.h>
 #include <math.h>
+#include <complex.h>
 #ifdef _MSC_VER
     #define int64_t signed __int64
     #define uint64_t unsigned __int64
     #define uint32_t unsigned __int32
+    #define _complex_float _Fcomplex
```
Comment (stuartarchibald) [inline numba/_helperlib.c:16]:
```suggestion
    #define _complex_float_t _Fcomplex
```
maybe name with `_t` suffix to indicate type?

File: numba/np/npyfuncs.py [hunk 3]
```diff
@@ -1590,23 +1638,39 @@ def np_real_nextafter_impl(context, builder, sig, args):
...
-        types.float64: 'npy_nextafter',
+        types.float32: 'numba_nextafterf',
+        types.float64: 'numba_nextafter',
     }
 
     return _dispatch_func_by_name_type(context, builder, sig, args,
                                        dispatch_table, 'nextafter')
 
 def np_real_spacing_impl(context, builder, sig, args):
+    # This is different to how NumPy does it, NumPy has a specialisation of
```
Comment (stuartarchibald) [inline numba/np/npyfuncs.py:1649]: Need to check that unit tests handle `inf` correctly.

Comment (stuartarchibald) [reply numba/np/npyfuncs.py:1649]: `inf` input appears to generally be untested in the ufunc tests, manually testing the `spacing` impl demonstrates that `spacing(np.inf)` returns `nan` regardless of the sign of the `np.inf`, this matches NumPy.

**Comment by seberg** [comment]
Ah, npy_math is distinct from the inner-loops (defines functions that end up being used there, often even via function pointers).  But I appreciate getting rid of it.  It is an akward library since it is currently statically linked, and we have to figure out what to do with it in NumPy.

**Comment by stuartarchibald** [comment]
> Ah, npy_math is distinct from the inner-loops (defines functions that end up being used there, often even via function pointers). But I appreciate getting rid of it. It is an akward library since it is currently statically linked, and we have to figure out what to do with it in NumPy.

I wasn't sure of the exact relationship between the two, perhaps this is something best explained/discussed at e.g. a Numba public meeting? This patch reduces reliance on an external (NumPy) implementation and also makes Numba's code in the areas changed optimisable by LLVM, am going to try and get this patch merged!

**Comment by stuartarchibald** [comment]
Doc build failing due to #8548.

File: numba/np/npyfuncs.py [hunk 4]
```diff
@@ -1590,23 +1638,39 @@ def np_real_nextafter_impl(context, builder, sig, args):
...
+    # translating this very involved function. Further, the NumPy comments note
+    # that there is a lot of redundancy present between the two.
     _check_arity_and_homogeneity(sig, args, 1)
 
     dispatch_table = {
-        types.float32: 'npy_spacingf',
-        types.float64: 'npy_spacing',
+        types.float32: 'numba_nextafterf',
+        types.float64: 'numba_nextafter',
     }
```
Comment (sklam) [inline numba/np/npyfuncs.py:1661]: `*sig.args, *sig.args` is an odd way to spell this. 

wouldn't something like below be more readable

```python
[ty] = sig.args
typing.signature(sig.return_type, ty, ty)
```

File: numba/np/npyfuncs.py (hunk 3 above)
Comment (sklam) [reply numba/np/npyfuncs.py:1649]: probably worth adding this test

File: numba/np/npyfuncs.py (hunk 4 above)
Comment (stuartarchibald) [reply numba/np/npyfuncs.py:1661]: Fixed in 7966e9b.

File: numba/np/npyfuncs.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/np/npyfuncs.py:1649]: Added in a28cf62.

**Comment by stuartarchibald** [comment]
Thanks for the review @sklam, comments should now be addressed.

**Comment by stuartarchibald** [comment]
Windows build is broken against a28cf62, I'll fix, but have also marked this as needing a full buildfarm run before merge.

**Review by sklam — DISMISSED** [review]
Thanks! LGTM. Just pending buildfarm.

**Comment by esc** [comment]
`Build numba_smoketest_cpu_yaml_151 has started`

**Comment by esc** [comment]
> `Build numba_smoketest_cpu_yaml_151 has started`

This was all green! 🍏

**Comment by sklam** [comment]
@stuartarchibald, there are conflicts

**Comment by stuartarchibald** [comment]
> @stuartarchibald, there are conflicts

@sklam thanks, should be fixed in c12b181.

**Comment by esc** [comment]
@stuartarchibald thank you for fixing the conflicts, I have re-scheduled this as:

`Build numba_smoketest_cpu_yaml_152 has started`

**Comment by esc** [comment]
> @stuartarchibald thank you for fixing the conflicts, I have re-scheduled this as:
> 
> `Build numba_smoketest_cpu_yaml_152 has started`

This never completed due to configuration issues on the CI system. So I started a new one at:

`Build numba_smoketest_cpu_yaml_154` has started

**Comment by stuartarchibald** [comment]
The test failures/issues in this patch due to strings `temp.c/cxx` appearing on stdout/stderr are due to https://github.com/numba/numba/pull/8671#issuecomment-1352937426. Fix is probably #8661.

**Comment by sklam** [comment]
`numba_smoketest_cpu_yaml_157`

**Comment by sklam** [comment]
> `numba_smoketest_cpu_yaml_157`

passed (except for known RNG issues)

**Review by sklam — APPROVED** [review]

**Pull request merged**